data-upgrade-users/src/main/java/org/exoplatform/migration/UserPasswordHashMigration.java

```diff
@@ -0,0 +1,147 @@
+          + " INNER JOIN jbid_io ON jbid_io_attr.IDENTITY_OBJECT_ID = jbid_io.ID)"
+          + " INNER JOIN( SELECT jbid_io_attr.IDENTITY_OBJECT_ID,"
+          + " min(CASE WHEN jbid_io_attr.NAME  = 'passwordSalt' THEN jbid_io_attr.NAME  ELSE NULL END) AS salt,"
+          + " min(CASE WHEN jbid_io_attr.NAME = 'passwordSalt128' THEN jbid_io_attr.NAME  ELSE NULL END) AS salt128"
+          + " FROM jbid_io_attr  GROUP BY jbid_io_attr.IDENTITY_OBJECT_ID"
+          + " HAVING salt IS NOT NULL AND salt128 IS NULL) jia ON jbid_io_attr.IDENTITY_OBJECT_ID  = jia.IDENTITY_OBJECT_ID;";
+
+      Query nativeQuery = entityManager.createNativeQuery(sqlString);
+      List<Object[]> result = nativeQuery.getResultList();
+      result.forEach(item -> {
+        String userName = (String) item[0];
+        String passwordHash = (String) item[1];
+        try {
+          String saltString = Hex.encodeHexString(generateRandomSalt());
+          User user = picketLinkIDMService.getIdentitySession().getPersistenceManager().findUser(userName);
+          picketLinkIDMService.getExtendedAttributeManager().addAttribute(userName, PASSWORD_SALT_USER_ATTRIBUTE, saltString);
+          picketLinkIDMService.getExtendedAttributeManager().updatePassword(user, passwordHash);
+          int count = updatedPasswords.getAndIncrement();
+          if (count % 50 == 0 || count == result.size()) {
+            LOG.info("{}/{} passwords have been updated", count, result.size());
+          }
+        } catch (Exception e) {
+          hasErrors = true;
+          LOG.error("Error while creating attribute salt and updating password hash for user : {}", userName, e);
+        }
+      });
+      if (hasErrors) {
+        throw new IllegalStateException("UserPasswordHashMigration upgrade failed due to previous errors");
+      }
+    } catch (Exception e) {
+      LOG.error("Error while getting old users passwords hash", e);
+      throw new IllegalStateException("UserPasswordHashMigration upgrade failed due to previous errors");
+    } finally {
+      RequestLifeCycle.end();
+    }
+    LOG.info("End upgrade of users passwords hashing algorithm. {} passwords has been updated. It took {} ms",
+             updatedPasswords.get(),
+             (System.currentTimeMillis() - startupTime));
+  }
+
+  @Override
+  public boolean shouldProceedToUpgrade(String newVersion,
+                                        String previousGroupVersion,
+                                        UpgradePluginExecutionContext previousUpgradePluginExecution) {
+    return getDefaultCredentialEncoderName().equals(DEFAULT_ENCODER)
+        && super.shouldProceedToUpgrade(newVersion, previousGroupVersion, previousUpgradePluginExecution);
+  }
+
+  private String getDefaultCredentialEncoderName() {
+    Map<String, String> props = new HashMap<>();
+    picketLinkIDMService.getConfigMD().getRealms().get(0).getOptions().forEach((k, v) -> props.put(k, v.get(0)));
```

Comment:
Can you explain this line ? 
There is twice get(0), are we ok with the fact that there is only one realm  ?
and that the option we search is really in first position ?

In addition, instead of creating a map<String, String>, could we use a filter/map operation to directly have the correct value to return ?

Reply:
It is the only way to do that, and it is one realm so it is in the first position,
 if you have another way could you suggest it ?

and i don't see how to use filter/map here and what difference will make.
Do you know that options is a map and not a list or array ? beside that the value of the map is another list.

Reply:
there is a solution to get default credential with proper way and is to declare a **public** function in `ExtendedAttributeManager.getDefaultCrdentialEncoder()`.
I didn't do that because i will need this function only in the UP

Reply:
I think that getDefaultCredential is clearer, and safer.
WDYT ?

Reply:
OK i will do it.